test/Streamly/Test/Prelude/Serial.hs

```diff
@@ -106,55 +107,258 @@ splitOnSuffixSeq = do
+            intercalateSplitEqIdNoSepEnd 1
+            intercalateSplitEqIdNoSepEndS 1
+            concatSplitIntercalateEqConcat splitOnSuffixSeq_ intercalateSuffix 1
+            concatSplitIntercalateEqConcatS
+                splitOnSuffixSeq_
+                IS.intercalateSuffix
+                1
+
+            -- Shift Or
+            intercalateSplitEqIdNoSepEnd 2
+            intercalateSplitEqIdNoSepEndS 2
+            concatSplitIntercalateEqConcat splitOnSuffixSeq_ intercalateSuffix 2
+            concatSplitIntercalateEqConcatS
+                splitOnSuffixSeq_
+                IS.intercalateSuffix
+                2
+
+            -- Karp-Rabin
+            intercalateSplitEqIdNoSepEnd 4
+            intercalateSplitEqIdNoSepEndS 4
+            concatSplitIntercalateEqConcat splitOnSuffixSeq_ intercalateSuffix 4
+            concatSplitIntercalateEqConcatS
+                splitOnSuffixSeq_
+                IS.intercalateSuffix
+                4
+
+            -- Exclusive case
+            splitIntercalateEqId splitOnSuffixSeq_ intercalateSuffix
+            splitIntercalateEqIdS splitOnSuffixSeq_ IS.intercalateSuffix
+
+    where
+
+    splitOnSeq_ xs ys =
+        S.toList $ IS.splitOnSeq (A.fromList ys) FL.toList (S.fromList xs)
+
+    splitOnSuffixSeq_ xs ys =
+        S.toList $ IS.splitOnSuffixSeq (A.fromList ys) FL.toList (S.fromList xs)
+
+    intercalateSuffix xs yss = intercalate xs yss ++ xs
+
+    nonSepElem :: Gen a
+    nonSepElem = suchThat arbitrary (/= sep)
+
+    listWithSep :: Gen [a]
+    listWithSep = listOf $ frequency [(3, arbitrary), (1, elements [sep])]
+
+    listWithoutSep :: Gen [a]
+    listWithoutSep = vectorOf 4 nonSepElem
+
+    listsWithoutSep :: Gen [[a]]
+    listsWithoutSep = listOf listWithoutSep
+
+    listsWithoutSep1 :: Gen [[a]]
+    listsWithoutSep1 = listOf1 listWithoutSep
+
+    intercalateSplitEqId i =
+        let name =
+                "intercalate . splitOnSeq == id ("
+                    <> show i <> " element separator)"
+         in prop name
+                $ forAll listWithSep
+                $ \xs -> do
+                      withMaxSuccess maxTestCount
+                          $ monadicIO
+                          $ do
+                              ys <- splitOnSeq_ xs (replicate i sep)
+                              listEquals
+                                  (==)
+                                  (intercalate (replicate i sep) ys)
+                                  xs
+
+    intercalateSplitEqIdS i =
+        let name =
+                "S.intercalate . splitOnSeq == id ("
+                    <> show i <> " element separator)"
+         in prop name
+                $ forAll listWithSep
+                $ \xs -> do
+                      withMaxSuccess maxTestCount
+                          $ monadicIO
+                          $ do
+                              ys <- splitOnSeq_ xs (replicate i sep)
+                              zs <-
```

Comment:
Why don't we move this step to `intercalateSplitEqId` itself and remove this function which is mostly the as the former. We can just use `listEquals xs ys` and `listEquals xs zs` in the same function.

Reply:
Yea, we can do that in all functions.

Reply:
I could probably do it for all of them but then it creates long lines. I've done it for the suggested